VCF batch ingestion

gspowley opened the pull request:
Add `tiledb.cloud.vcf.ingest` to support ingestion of VCF files.

Example:
```python
from tiledb.cloud.vcf import ingest

dataset_uri = "s3://..."
config = {}
conf['vfs.s3.access_key_id'] = ...
conf['vfs.s3.secret_access_key'] = ...
search_uri = "s3://1000genomes-dragen-v3.7.6/data/individuals/hg38-graph-based"
pattern = "*.hard-filtered.gvcf.gz"

# limit ingestion size for testing
max_files=20

ingest(
    dataset_uri,
    config=config,
    search_uri=search_uri,
    pattern=pattern,
    max_files=max_files,
)

```

ktsitsi commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,1107 @@
+# --------------------------------------------------------------------
+
+
+def read_config(path: str = "~/.aws/credentials"):
+    """Read config values from a file."""
+
+    config = {}
+
+    cp = configparser.ConfigParser()
+    cp.read(os.path.expanduser(path))
+    try:
+        config["vfs.s3.aws_access_key_id"] = cp["default"]["aws_access_key_id"]
+        config["vfs.s3.aws_secret_access_key"] = cp["default"]["aws_secret_access_key"]
+    except KeyError:
+        pass
+
+    return config
```
These `utils` can be used in other ingestion functions in the future, but also in some currently implemented e.g. `bioimg`. Could we move it under a `utils` folder for common use?

gspowley replied:
Yes, it's a good idea to share the utility functions and the event logger. I can move them to `tiledb.cloud.utils` or somewhere else, if there's a better place.

gspowley replied:
Reusable functions are now in `tiledb.cloud.utilities` since `utils.py` already existed.

thetorpedodog replied:
> Reusable functions are now in `tiledb.cloud.utilities` since `utils.py` already existed.

`utils` is really for internal consumption only; it could be moved into the `_common` directory.

Shelnutt2 commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,987 @@
+
+    logger = get_logger()
+
+    graph = tiledb.cloud.dag.DAG(
```
This is where I think we should be using a dynamic dag instead of running three separate dags. @NikolaosPapailiou Can you work with @gspowley and show this could be switched to a dynamic data for determining the files to ingest.

NikolaosPapailiou replied:
Sure I will be happy to help with this. The main idea would be to have only one DAG that does all the processing (generate_uris, update_manifest, ingest). Dynamic taskgraphs can help with dynamically creating tasks based on the output of other task. So in this example generate_uris could return a list of sample uris that update_manifest and ingest tasks expand upon creating one task per sample uri. This can be done by just returning a list object and adding `expand_output` on the dynamic DAG node (you can check an example in https://github.com/TileDB-Inc/TileDB-Cloud-Py/blob/main/tests/test_dag.py#L643).

Please keep in mind that the `expand_output` functionality is only available for bath taskgraphs. So if you want the same logic to run with realtime taskgraphs the code needs to be duplicated and follow another pattern for local or realtime execution.

gspowley commented:
VCF ingestion is ready for review and release as an experimental feature while we continue to make improvements.

Shelnutt2 commented on src/tiledb/cloud/utilities/common.py:
```diff
@@ -0,0 +1,196 @@
+import tiledb.cloud
+
+
+def read_config(
```
Can we rename this to something like `read_aws_config`?

Shelnutt2 commented on src/tiledb/cloud/utilities/common.py:
```diff
@@ -0,0 +1,196 @@
+    return config
+
+
+def set_context(config: Mapping[str, Any] = {}):
```
Lets rename this to `set_aws_context`, or support more cloud vendors. Is there a reason this is setting os envs?

gspowley replied:
We can rename this now and support more cloud vendors in the future. It's setting os envs for commands that access S3 directly, like AWS SDK CLI and bcftools.

Shelnutt2 reviewed (approved):
Some minor nits, we need to get some follows tracking better handling for the configs. This can be all done in follow ups.

thetorpedodog commented on src/tiledb/cloud/utilities/common.py:
```diff
@@ -0,0 +1,197 @@
+import tiledb.cloud
+
+
+def read_aws_config(
```
Does this need to be exported for general use? It seems like it is mostly useful as an internal tool for other functions.

It's fine if we want to; I just want to make sure we keep a sensible API surface.

gspowley replied:
This is useful in a notebook environment, where we have to read the credentials from somewhere. We have been passing around functions in notebooks that do the same thing, so I thought we could formalize it.

It's fine if we take some time to think about it. Let me know and I can remove `read_aws_config`.

thetorpedodog commented on src/tiledb/cloud/utilities/common.py:
```diff
@@ -0,0 +1,197 @@
+    return config
+
+
+def set_aws_context(config: Mapping[str, Any] = {}):
```
While it looks like `config` is never mutated, I'm still uneasy with mutable objects as defaults. Maybe just `None`?

Also, add `-> None` to these functions which return nothing.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+from tiledb.cloud.utilities import set_aws_context
+from tiledb.cloud.utilities import write_log_event
+
+LOCAL_INGEST = False
```
Since these aren't really constants they probably shouldn't be uppercase.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+TRACE = False
+
+
+def local_ingest():
```
I don't entirely understand the intent of these functions. Could a user not simply set the module-level variable themselves?

If you want to keep them, it would be helpful to give them a name that reflects that they are taking an action, maybe `set_local_ingest`.

gspowley replied:
These functions can go away. They were added with some other code to test the code on Cloud without putting everything in one file, and they probable weren't necessary then either.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+
+
+def setup(
+    config: Union[Mapping[str, Any], None],
```
Optional is exactly equivalent to Union[..., None].

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+    config: Union[Mapping[str, Any], None],
+    dataset_uri: Optional[str] = None,
+    id: Optional[str] = None,
+) -> Union[logging.Logger, Tuple[logging.Logger, Profiler]]:
```
It's not obvious from the call under what circumstances the function will return a logger versus a logger AND a profiler.

To avoid a potential pitfall for users when using this function, I would instead make the return type one shape:

`-> Tuple[logging.Logger, Optional[Profiler]:`.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+MANIFEST_ARRAY = "manifest"
+
+# Default attributes to materialize
+DEFAULT_ATTRIBUTES = ["fmt_GT", "fmt_DP", "fmt_GQ"]
```
This should probably be a tuple.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+    d0 = tiledb.Dim(name="sample_name", dtype="ascii", filters=ascii_fl)
+    dom = tiledb.Domain(d0)
+
+    attrs = []
+    attrs += [tiledb.Attr(name="status", dtype="ascii", filters=ascii_fl)]
+    attrs += [tiledb.Attr(name="vcf_uri", dtype="ascii", filters=ascii_fl)]
+    attrs += [tiledb.Attr(name="vcf_bytes", dtype=np.uint64, filters=int_fl)]
+    attrs += [tiledb.Attr(name="index_uri", dtype="ascii", filters=ascii_fl)]
+    attrs += [tiledb.Attr(name="index_bytes", dtype=np.uint64, filters=int_fl)]
+    attrs += [tiledb.Attr(name="records", dtype=np.uint64, filters=int_fl)]
```
```python
attrs = [
    tiledb.Attr(...),
    ...,
]
```
?

thetorpedodog commented on src/tiledb/cloud/utilities/profiler.py:
```diff
@@ -0,0 +1,194 @@
+
+        # Log node start time and system info
+        node_id = open("/proc/sys/kernel/random/boot_id").read().strip()
+        cmd = "uptime && free -h && uname -a && lscpu"
```
Would it be better to run these individually so that (a) we're not using `shell=True` and (b) we get the data back in a more structured manner?

thetorpedodog commented on src/tiledb/cloud/utilities/profiler.py:
```diff
@@ -0,0 +1,194 @@
+        :return: stdout from the command
+        """
+
+        res = subprocess.run(cmd, shell=True, capture_output=True, text=True)
+        if res.stderr:
+            print(res.stderr)
```
`stdout=subprocess.PIPE, stderr=subprocess.STDERR`?

thetorpedodog commented on src/tiledb/cloud/utilities/profiler.py:
```diff
@@ -0,0 +1,194 @@
+        self.stats = ["time,cpu,mem"]
+        self._timeout()
+
+    def run_cmd(self, cmd: str) -> str:
```
This doesn't use any self-state. Should it be a module-level function?

thetorpedodog commented on src/tiledb/cloud/utilities/profiler.py:
```diff
@@ -0,0 +1,194 @@
+        }
+
+
+class Profiler(object):
```
It seems like it might make more sense to have this as a context manager so you can do:

```python
with my_profiler.profile(whatever_param):
  ...
```

This might also help with the potential dangling array below (`self.array = tiledb.open(uri, "w")`)

thetorpedodog commented on src/tiledb/cloud/utilities/profiler.py:
```diff
@@ -0,0 +1,194 @@
+        """
+
+        try:
+            cpu = open("/sys/fs/cgroup/cpu/cpuacct.usage").readline().strip()
```
consider `pathlib.Path(...).read_text().strip()` to ensure everything is closed (and maybe just pull open-read-strip into a function)

thetorpedodog commented on src/tiledb/cloud/utilities/common.py:
```diff
@@ -0,0 +1,197 @@
+    print_logs(graph, debug)
+
+
+def print_logs(graph: tiledb.cloud.dag.DAG, debug: bool = False):
```
`debug` should probably be kw-only.

thetorpedodog commented on src/tiledb/cloud/utilities/profiler.py:
```diff
@@ -0,0 +1,194 @@
+        self,
+        uri: str,
+        id: str,
+        trace: bool = False,
```
kw-only here too

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+import numpy as np
+import tiledbvcf
+
+import tiledb.cloud
```
To avoid circular import weirdness, it would be best to import the specific modules that contain whatever you're using.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+    level = logging.DEBUG if VERBOSE else logging.INFO
+    logger = get_logger(__name__, level)
+
+    logger.debug(
```
let the logger do the substitution:

```python
logger.debug(
    "tiledb=%s, libtiledb=%s, tiledbvcf=%s", tiledb.version(), ...
)

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+
+    # Check if the dataset already exists
+    if tiledb.object_type(dataset_uri) != "group":
+        logger.info(f"Creating dataset: '{dataset_uri}'")
```
e.g. here `logger.info("Creating dataset: %r", dataset_uri)`

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+    use_s3 = search_uri.startswith("s3://")
+    # Run command to find URIs matching the pattern
+    if use_s3:
+        cmd = f"exec aws s3 sync --dryrun --exclude '*' --include '{pattern}' {search_uri} ."
```
Using `shell` here is vulnerable to accidental shell injection; instead the command should be built as a sequence:

```python
cmd = ("aws", "s3", "sync" "--dryrun", "--exclude", "*", "--include", pattern, search_uri, ".")
```

and then executed without `shell=True`.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+        stdout=subprocess.PIPE,
+    )
+
+    # Optionally ignore URIs and limit the number returned
+    if ignore:
+        cmd = f"grep -Ev '{ignore}'"
+    else:
+        cmd = "grep ''"
+    if max_files:
+        cmd += f" -m {max_files}"
+
+    logger.debug(cmd)
+    p2 = subprocess.Popen(
+        cmd,
+        shell=True,
+        text=True,
+        stdin=p1.stdout,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
```
Could we do this in Python, iteratively reading in lines and handling them as needed rather than shelling out to grep?

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,977 @@
+    # (dryrun) download: s3://1000genomes-dragen-v3.7.6/foo to foo
+    result = []
+    if res_stdout:
+        for line in res_stdout.strip().split("\n"):
```
.splitlines?

gspowley commented:
Thanks for the detailed review @thetorpedodog! I addressed all of the comments except the ones related to `find_uris_udf`, which we plan to rewrite and replace in a future PR.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,970 @@
+    num_partitions = ceil(len(sample_uris) / batch_size)
+    num_consolidates = ceil(num_partitions / workers)
+
+    consolidate = None
+    for i in range(num_partitions):
+        if i % workers == 0:
+            prev_consolidate = consolidate
+            consolidate = submit(
+                consolidate_dataset_udf,
+                dataset_uri,
+                config=config,
+                exclude=None,
+                include=[MANIFEST_ARRAY, LOG_ARRAY],
+                id=f"manifest-consol-{i//workers}",
+                name=f"Consolidate VCF Manifest {i//workers + 1}/{num_consolidates} ",
+            )
+
+        ingest = submit(
+            ingest_manifest_udf,
+            dataset_uri,
+            sample_uris[i * batch_size : (i + 1) * batch_size],
+            config=config,
+            id=f"manifest-ingest-{i}",
+            name=f"Ingest VCF Manifest {i+1}/{num_partitions} ",
+        )
+        if prev_consolidate:
+            ingest.depends_on(prev_consolidate)
+
+        if consolidate:
+            consolidate.depends_on(ingest)
```
This is a little hard for me to follow. Is this designed to have N ingest steps, then a consolidate step, then N more ingest steps, then a consolidate step, etc.? It might be worth adding a comment explaining this.

gspowley replied:
Thanks again @thetorpedodog. I addressed the latest comments and got rid of all the `shell=True` subprocesses.

thetorpedodog reviewed (commented):
Overall looks pretty good. A few things I didn't see last time but mostly minor changes now.

thetorpedodog commented on src/tiledb/cloud/vcf/ingestion.py:
```diff
@@ -0,0 +1,997 @@
+            return 0
+
+            cmd = ["bcftools", "index", "-n", vcf]
+            res = subprocess.run(cmd, capture_output=True, text=True)
+            if res.stderr:
+                print(res.stderr)
+                return 0
```
This code won’t execute, so this is not a critical thing, but:

Just to make sure—if there is any stderr output, you want to return 0? If that’s what you want to do, that’s fine, but it might be worth adding a comment explaining why.

gspowley replied:
Yes, the 0 in the manifest will indicate there's a problem with the index that needs to be addressed. Added a comment there to explain.

thetorpedodog reviewed (approved):
One suggestion but otherwise this looks good!
